runtime/executor/src/main/java/org/apache/nemo/runtime/executor/task/MultiThreadParentTaskDataFetcher.java

```diff
@@ -96,14 +100,37 @@ private void fetchDataLazily() {
     final List<CompletableFuture<DataUtil.IteratorWithNumBytes>> futures = readersForParentTask.read();
     numOfIterators = futures.size();
 
+    if (numOfIterators > 1) {
+      inputWatermarkManager = new MultiInputWatermarkManager(numOfIterators, new WatermarkCollector());
+    } else {
+      inputWatermarkManager = new SingleInputWatermarkManager(new WatermarkCollector());
+    }
+
     futures.forEach(compFuture -> compFuture.whenComplete((iterator, exception) -> {
       // A thread for each iterator
       queueInsertionThreads.submit(() -> {
         if (exception == null) {
           // Consume this iterator to the end.
           while (iterator.hasNext()) { // blocked on the iterator.
             final Object element = iterator.next();
-            elementQueue.offer(element);
+
+            if (LOG.isDebugEnabled()) {
+              LOG.debug("Receive data : {}", element);
+            }
+
+            if (element instanceof WatermarkWithIndex) {
+              // watermark element
+              // the input watermark manager is accessed by multiple threads
+              // so we should synchronize it
+              synchronized (inputWatermarkManager) {
+                final WatermarkWithIndex watermarkWithIndex = (WatermarkWithIndex) element;
+                inputWatermarkManager.trackAndEmitWatermarks(
```

Comment:
make the method `trackAndEmitWatermarks()` synchronized to always guarantee synchronization?

Reply:
will add the comment to `InputWatermarkManager`